### ISSUE TITLE
WARNING: Unexpected CBForest exception iterating query

### ISSUE THREAD

**Issue by brendand** [couchbase/couchbase-lite-ios#899]
In reference to my post on Google Forums, https://groups.google.com/forum/#!topic/mobile-couchbase/DFObsy14aEk I'm receiving the following error:

WARNING: Unexpected CBForest exception iterating query

I believe it's occurring when my map/reduce block is being executed as it seems to occur every time I change things in my app which would cause the CBLView to be re-generated.

In the SQLite version of my database for a particular form, I'm getting 11,040 records, but in the ForestDB version I'm getting 769 records, which corresponds to the number of rows in maps_43 table in SQLite. Unfortunately I can't read the corresponding ForestDB viewIndex in any legible way, but in my database file, I believe it corresponds to the records-frm-6F712B3923A14F8D9D3388CE2626A962.viewindex file.

The only change I made to my app is to change the storageType to ForestDB and then I re-migrated my records from my old database.

I'd be happy to send the ForestDB and SQLite versions of my database to someone, but I don't want to post a public link to it since it has some customer data in it and it's also from an unreleased version of my app.

I'm using commit: 4ded1caaa92db5a4cd177ad52f8115d2c7a2ea0e from master on Mac OS X 10.11 Beta (15A263e)

**Comment by snej** [comment]
Just to make sure I understand: when you talk about the unexpected number of records, you mean the number of rows in a query result, not the total number of documents in the database, right?

I think the best place to start is by adding a breakpoint on C++ exceptions. (If you haven't done this before: go to the Breakpoints navigator, press the + button at the bottom, choose "Add Exception Breakpoint" from the pop-up.)

Now repeat the action that causes the warning; you should drop into the debugger with a C++ exception. Capture the backtrace with a `bt` command to the debugger, and post it here.

**Comment by brendand** [comment]
Hi Jens,

Yes, I meant the number of rows in a query result.

Here's a backtrace:

```
(lldb) bt
CouchbaseLite was compiled with optimization - stepping may behave oddly; variables may not be available.
* thread #1: tid = 0x1eb28c, 0x00007fff9d9b6b99 libc++abi.dylib`__cxa_throw, queue = 'com.apple.main-thread', stop reason = breakpoint 16.1 17.1
    frame #0: 0x00007fff9d9b6b99 libc++abi.dylib`__cxa_throw
  * frame #1: 0x0000000100b6a0e0 CouchbaseLite`forestdb::CollatableReader::readString(forestdb::CollatableTypes::Tag) [inlined] forestdb::CollatableReader::expectTag(this=<unavailable>) + 256 at Collatable.cc:135 [opt]
    frame #2: 0x0000000100b6a0cc CouchbaseLite`forestdb::CollatableReader::readString(this=<unavailable>, tag=<unavailable>) + 236 at Collatable.cc:168 [opt]
    frame #3: 0x0000000100b69fd3 CouchbaseLite`forestdb::CollatableReader::readString(this=<unavailable>) + 19 at Collatable.cc:160 [opt]
    frame #4: 0x0000000100b6f73c CouchbaseLite`forestdb::IndexEnumerator::read(this=0x000000011620cec0) + 412 at Index.cc:314 [opt]
    frame #5: 0x0000000100b1e534 CouchbaseLite`__58-[CBL_ForestDBViewStorage regularQueryWithOptions:status:]_block_invoke(.block_descriptor=0x0000600002492660) + 1630 at CBL_ForestDBViewStorage.mm:573 [opt]
    frame #6: 0x0000000100b397ef CouchbaseLite`-[CBLQueryEnumerator nextRow](self=0x0000600002762dc0, _cmd=<unavailable>) + 140 at CBLQueryEnumerator.m:156 [opt]
    frame #7: 0x0000000100b39a63 CouchbaseLite`-[CBLQueryEnumerator countByEnumeratingWithState:objects:count:](self=<unavailable>, _cmd=<unavailable>, state=0x00007fff5fbfaf20, stackbuf=<unavailable>, stackbufLength=16) + 117 at CBLQueryEnumerator.m:211 [opt]
    frame #8: 0x00007fff9ed1029f CoreFoundation`-[NSEnumerator allObjects] + 287
    frame #9: 0x0000000100b39899 CouchbaseLite`-[CBLQueryEnumerator allObjects](self=0x0000600002762dc0, _cmd=<unavailable>) + 73 at CBLQueryEnumerator.m:174 [opt]
    frame #10: 0x0000000100b39463 CouchbaseLite`-[CBLQueryEnumerator sortUsingDescriptors:](self=<unavailable>, _cmd=<unavailable>, sortDescriptors=<unavailable>) + 163 at CBLQueryEnumerator.m:126 [opt]
    frame #11: 0x0000000100b396be CouchbaseLite`-[CBLQueryEnumerator sortUsingDescriptors:skip:limit:](self=0x0000600002762dc0, _cmd=<unavailable>, sortDescriptors=<unavailable>, skip=0, limit=4294967295) + 40 at CBLQueryEnumerator.m:133 [opt]
    frame #12: 0x0000000100af1db5 CouchbaseLite`-[CBLQuery run:](self=0x0000600000b8b530, _cmd=<unavailable>, outError=domain: class name = NSKVONotifying_TFForm - code: 105553236080512) + 332 at CBLQuery.m:232 [opt]
    frame #13: 0x000000010007de68 Tap Forms`-[TFForm recordsGroupedBySection](self=0x0000000100f619f0, _cmd="recordsGroupedBySection") + 3080 at TFForm.m:565
    frame #14: 0x000000010014a1f6 Tap Forms`-[FormEntryListViewController refreshFormEntries](self=0x000000011546f550, _cmd="refreshFormEntries") + 5286 at FormEntryListViewController.m:730
    frame #15: 0x000000010014c166 Tap Forms`-[FormEntryListViewController formSelectionChanged:](self=0x000000011546f550, _cmd="formSelectionChanged:", selection=0x00006080004a1b60) + 3622 at FormEntryListViewController.m:889
    frame #16: 0x00000001002124d7 Tap Forms`-[SelectedFormSplitViewController handleFormChanged:](self=0x000000011545d450, _cmd="handleFormChanged:", notification=@"TFFormSelectionDidChangeNotification") + 1319 at SelectedFormSplitViewController.m:534
    frame #17: 0x00007fff9ed766bc CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    frame #18: 0x00007fff9ed7662f CoreFoundation`___CFXRegistrationPost_block_invoke + 63
    frame #19: 0x00007fff9ed75cf7 CoreFoundation`_CFXRegistrationPost + 407
    frame #20: 0x00007fff9ed75a62 CoreFoundation`___CFXNotificationPost_block_invoke + 50
    frame #21: 0x00007fff9ed6fcf2 CoreFoundation`-[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1922
    frame #22: 0x00007fff9ec5e845 CoreFoundation`_CFXNotificationPost + 693
    frame #23: 0x00007fff9ff888d1 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 66
    frame #24: 0x00000001002b8b7d Tap Forms`-[MainWindowController _switchRightContentViewFromController:](self=0x0000000102651f10, _cmd="_switchRightContentViewFromController:", controllerToUse=0x000000011545d450) + 2269 at MainWindowController.m:842
    frame #25: 0x00000001002b954f Tap Forms`-[MainWindowController _chooseRightControllerBasedOnSelection](self=0x0000000102651f10, _cmd="_chooseRightControllerBasedOnSelection") + 2383 at MainWindowController.m:888
    frame #26: 0x00000001002b7f65 Tap Forms`-[MainWindowController _selectionWillChange:](self=0x0000000102651f10, _cmd="_selectionWillChange:", not=@"TFFormSelectionWillChangeNotification") + 4629 at MainWindowController.m:767
    frame #27: 0x00007fff9ed766bc CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    frame #28: 0x00007fff9ed7662f CoreFoundation`___CFXRegistrationPost_block_invoke + 63
    frame #29: 0x00007fff9ed75cf7 CoreFoundation`_CFXRegistrationPost + 407
    frame #30: 0x00007fff9ed75a62 CoreFoundation`___CFXNotificationPost_block_invoke + 50
    frame #31: 0x00007fff9ed6fcf2 CoreFoundation`-[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1922
    frame #32: 0x00007fff9ec5e845 CoreFoundation`_CFXNotificationPost + 693
    frame #33: 0x00007fff9ff888d1 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 66
    frame #34: 0x00000001003eb27c Tap Forms`-[SourceListViewController outlineViewSelectionDidChange:](self=0x0000000100f49a90, _cmd="outlineViewSelectionDidChange:", notification=@"NSOutlineViewSelectionDidChangeNotification") + 4316 at SourceListViewController.m:688
    frame #35: 0x00007fff9ed766bc CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    frame #36: 0x00007fff9ed7662f CoreFoundation`___CFXRegistrationPost_block_invoke + 63
    frame #37: 0x00007fff9ed75cf7 CoreFoundation`_CFXRegistrationPost + 407
    frame #38: 0x00007fff9ed75a62 CoreFoundation`___CFXNotificationPost_block_invoke + 50
    frame #39: 0x00007fff9ed6fcf2 CoreFoundation`-[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1922
    frame #40: 0x00007fff9ec5e845 CoreFoundation`_CFXNotificationPost + 693
    frame #41: 0x00007fff9ff888d1 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 66
    frame #42: 0x00007fff927ad5b3 AppKit`-[NSTableView _sendSelectionChangedNotificationForRows:columns:] + 175
    frame #43: 0x00007fff9277e78a AppKit`-[NSTableView _enableSelectionPostingAndPost] + 424
    frame #44: 0x00007fff92906bc8 AppKit`-[NSTableView mouseDown:] + 5333
    frame #45: 0x00007fff929054ba AppKit`-[NSOutlineView mouseDown:] + 74
    frame #46: 0x00007fff92e07b5d AppKit`-[NSWindow _handleMouseDownEvent:isDelayedEvent:] + 6322
    frame #47: 0x00007fff92e08ded AppKit`-[NSWindow _reallySendEvent:isDelayedEvent:] + 212
    frame #48: 0x00007fff92762795 AppKit`-[NSWindow sendEvent:] + 517
    frame #49: 0x00007fff9275eeb7 AppKit`-[NSApplication sendEvent:] + 2540
    frame #50: 0x00007fff9269212a AppKit`-[NSApplication run] + 796
    frame #51: 0x00007fff92614396 AppKit`NSApplicationMain + 1176
    frame #52: 0x0000000100353d02 Tap Forms`main(argc=3, argv=0x00007fff5fbff768) + 34 at main.m:15
    frame #53: 0x00007fff8e9ec5ad libdyld.dylib`start + 1
    frame #54: 0x00007fff8e9ec5ad libdyld.dylib`start + 1
```

When the breakpoint was fired, it stopped in this method:

```
    void CollatableReader::expectTag(Tag tag) {
        slice tagSlice = _data.read(1);
        if (tagSlice.size == 0)
            throw "unexpected end of collatable data";
        else if (tagSlice[0] != tag)
            throw "unexpected tag";
    }
```

on the `throw "unexpected end of collectable data";` line

**Comment by snej** [comment]
What kind of keys are you emitting into this view index? And what are the query's attributes?

**Comment by brendand** [comment]
Here's an example of some of the keys of this particular view:

```
["SANDRA"]
["MARTIN"]
["AGUSTIN ANDRÉS"]
```

If the form was sorted by more than one field you would have multiple keys in the array. But this particular one is only sorted by the First Name field. 

After I create my view, I simply call

```
CBLQuery *recordsQuery = [self recordsQuery];
recordsQuery.sortDescriptors = sortDescriptors;
recordsQuery.mapOnly = YES;
recordsQuery.prefetch = YES;
```

The recordsQuery method creates the view and then calls `[recordsView createQuery]` and then returns that.

The view is created like this:

```
NSString *viewName = [NSString stringWithFormat:@"records-%@", self.document.documentID];
CBLView *recordsView = [self.database existingViewNamed:viewName];
if (!recordsView) {
	recordsView = [self.database viewNamed:viewName];
}
recordsView.documentType = self.document.documentID;

[recordsView setMapBlock: MAPBLOCK({
	NSDictionary *valuesDict = doc[@"values"];
	NSMutableArray *values = [NSMutableArray array];

... lots of code to determine what key to emit ...

	if (values.count > 0) {
		emit(values, nil);
	}
		
}) reduceBlock:^id(NSArray *keys, NSArray *values, BOOL rereduce) {
	return @(values.count);
} version: versionNumber];
```

I'm actually not using the reduceBlock for anything in particular since we can now get the count from totalRows, so just ignore that and I don't think it has any significance to this issue.

**Comment by snej** [comment]
Something looks wrong with that index. Could you email me the database directory so I can examine it in detail? (Or just the file for that view index?)

**Comment by brendand** [comment]
Sure, what email address?

**Comment by snej** [comment]
Oh wow, nasty bug. Looks like a handful of the strings in your keys end with control characters (other than tab, newline or return). For arcane reasons the view index stores keys in a non-ASCII encoding, but I forgot to handle all control characters in that conversion, so the unhandled ones map to zero bytes … which are also used in the encoding to mean end-of-string. That causes parsing to break when reading the components of the key.

I have a fix I'll check in soon. You'll need to rebuild the indexes, though.

**Comment by snej** [comment]
Actually the indexes will rebuild themselves; I bumped an internal version number so it'll detect the out of date index and delete it.

**Comment by brendand** [comment]
That's fantastic Jens! And it worked!

However this build doesn't compile for Mac. You must have warnings set to be errors. I got a couple errors building `GTMNSData+zlib.m`. Just a couple of precision error:

    unsigned gotBack = kChunkSize - strm.avail_out;

I think should be

    uLong gotBack = kChunkSize - strm.avail_out;

and:

     _GTMDevAssert(strm.avail_in == 0,
                @"thought we finished deflate w/o using all input, %u bytes left",
                strm.avail_in);

I corrected them by changing to:

    uLong gotBack = kChunkSize - strm.avail_out;

and

    _GTMDevLog(@"thought we finished inflate w/o using all input, %lu bytes left",
               strm.avail_in);

That seems to have done the trick.

Basically changing `unsigned` to `uLong` and `%u` to `%lu`.

Thanks!

Brendan

**Comment by snej** [comment]
That's strange; I was building the Mac target, and our built-bots always build both Mac and iOS. And that source file hasn't been modified since 2012.

What target are you building, exactly? And what version of Xcode?

**Comment by brendand** [comment]
Xcode Version 7.0 beta 6 (7A192o) on El Capitan.

I'm building the CBL Listener Mac since it includes both the Listener and the CouchbaseLite frameworks.

Oh strange. I think Xcode 7 let me modify the zlib.h file from /usr/include. I wouldn't have thought that I could modify that file without entering in a root password. But it seems like it does. And I had modified that file for my own project which was throwing some warnings about precision. I didn't pay attention and assumed it was a file inside my own project. Never mind then.
